File: iOS/CityWalk/CityWalk/Models/Route.swift
import Foundation
import CoreLocation

// MARK: - 动态 CodingKey（用于兼容 "_id" / "id" 两种字段名）
private struct DynamicCodingKey: CodingKey {
    var stringValue: String
    var intValue: Int? { nil }
    init?(stringValue: String) { self.stringValue = stringValue }
    init?(intValue: Int) { nil }
}

// MARK: - 路线模型
struct Route: Identifiable, Codable, Hashable {
    let id: String
    let name: String
    let description: String?
    let distance: Double  // 米
    let duration: Int?     // 秒
    let difficulty: Difficulty?
    let tags: [String]?
    let coverImage: String?
    let points: [RoutePoint]?
    let pois: [POI]?
    let createdAt: Date?
    let createdBy: String?
    let city: String?
    let district: String?
    let elevationGain: Double?
    let favoritesCount: Int?
    let viewsCount: Int?
    let completionsCount: Int?
    let isPublished: Bool?
    let score: Double?
    let isFeatured: Bool?

    enum CodingKeys: String, CodingKey {
        case id = "_id"
        case name, description, distance, difficulty, tags, city, district, score
        case duration = "estimated_duration"
        case coverImage = "cover"
        case points, pois
        case createdAt = "created_at"
        case createdBy = "created_by"
        case elevationGain = "elevation_gain"
        case favoritesCount = "favorites_count"
        case viewsCount = "views_count"
        case completionsCount = "completions_count"
        case isPublished = "is_published"
        case isFeatured = "is_featured"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        // 兼容 "_id" 和 "id" 两种格式：后端 RouteDetail.model_dump() 输出 "id"，MongoDB 原始数据用 "_id"
        if let underscoreId = try? container.decodeIfPresent(String.self, forKey: .id) {
            // CodingKeys.id = "_id"，从 JSON 的 "_id" 读取
            id = underscoreId
        } else {
            // fallback: 从 JSON 的 "id" 读取（使用动态 key）
            let anyContainer = try decoder.container(keyedBy: DynamicCodingKey.self)
            id = (try? anyContainer.decodeIfPresent(String.self, forKey: DynamicCodingKey(stringValue: "id")!)) ?? ""
        }
        name = try container.decode(String.self, forKey: .name)
        description = try container.decodeIfPresent(String.self, forKey: .description)
        distance = try container.decode(Double.self, forKey: .distance)
        duration = try container.decodeIfPresent(Int.self, forKey: .duration)
        difficulty = try container.decodeIfPresent(Difficulty.self, forKey: .difficulty)
        tags = try container.decodeIfPresent([String].self, forKey: .tags)
        coverImage = try container.decodeIfPresent(String.self, forKey: .coverImage)
        points = try container.decodeIfPresent([RoutePoint].self, forKey: .points)
        pois = try container.decodeIfPresent([POI].self, forKey: .pois)
        createdAt = try container.decodeIfPresent(Date.self, forKey: .createdAt)
        createdBy = try container.decodeIfPresent(String.self, forKey: .createdBy)
        city = try container.decodeIfPresent(String.self, forKey: .city)
        district = try container.decodeIfPresent(String.self, forKey: .district)
        elevationGain = try container.decodeIfPresent(Double.self, forKey: .elevationGain)
        favoritesCount = try container.decodeIfPresent(Int.self, forKey: .favoritesCount)
        viewsCount = try container.decodeIfPresent(Int.self, forKey: .viewsCount)
        completionsCount = try container.decodeIfPresent(Int.self, forKey: .completionsCount)
        isPublished = try container.decodeIfPresent(Bool.self, forKey: .isPublished)
        score = try container.decodeIfPresent(Double.self, forKey: .score)
        isFeatured = try container.decodeIfPresent(Bool.self, forKey: .isFeatured)
    }
    
    // 格式化距离显示
    var formattedDistance: String {
        if distance >= 1000 {
            return String(format: "%.1f km", distance / 1000)
        } else {
            return String(format: "%.0f m", distance)
        }
    }
    
    // 格式化时长显示
    var formattedDuration: String {
        guard let duration = duration else { return "未知" }
        let hours = duration / 3600
        let minutes = (duration % 3600) / 60
        if hours > 0 {
            return "\(hours)小时\(minutes)分钟"
        } else {
            return "\(minutes)分钟"
        }
    }
}

// MARK: - 难度等级
enum Difficulty: String, Codable, Hashable {
    case easy = "easy"
    case medium = "medium"
    case hard = "hard"
    
    // 显示文本
    var displayText: String {
        switch self {
        case .easy: return "简单"
        case .medium: return "中等"
        case .hard: return "困难"
        }
    }
    
    var color: String {
        switch self {
        case .easy: return "green"
        case .medium: return "orange"
        case .hard: return "red"
        }
    }
    
    var icon: String {
        switch self {
        case .easy: return "leaf.fill"
        case .medium: return "flame.fill"
        case .hard: return "mountain.2.fill"
        }
    }
}

// MARK: - 轨迹点
struct RoutePoint: Codable, Hashable {
    let location: Location
    let elevation: Double?
    let timestamp: Date?
    let isWaypoint: Bool?
    let photos: [String]?
    let isEdited: Bool?
    let name: String?
    let description: String?
    
    enum CodingKeys: String, CodingKey {
        case location, elevation, timestamp, photos, name, description
        case isWaypoint = "is_waypoint"
        case isEdited = "is_edited"
    }
}

// MARK: - 位置
struct Location: Codable, Hashable {
    let type: String
    let coordinates: [Double]  // [经度, 纬度]
    
    var longitude: Double { coordinates.count > 0 ? coordinates[0] : 0 }
    var latitude: Double { coordinates.count > 1 ? coordinates[1] : 0 }
    
    var coordinate: CLLocationCoordinate2D {
        CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }
}

// MARK: - POI
struct POI: Identifiable, Codable, Hashable {
    let id: String
    let name: String
    let location: Location
    let category: String?
    let poiDescription: String?
    let images: [String]?
    let rating: Double?
    let tags: [String]?
    let amapPoiId: String?
    let distance: Double?
    let type: String?
    let address: String?
    
    enum CodingKeys: String, CodingKey {
        case id, name, location, category, images, rating, tags, distance, type, address
        case poiDescription = "description"
        case amapPoiId = "amap_poi_id"
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        id = try container.decode(String.self, forKey: .id)
        name = try container.decode(String.self, forKey: .name)
        location = try container.decode(Location.self, forKey: .location)
        category = try container.decodeIfPresent(String.self, forKey: .category)
        // description 可能是 String 或其他类型
        if let descString = try? container.decode(String.self, forKey: .poiDescription) {
            poiDescription = descString
        } else {
            poiDescription = nil
        }
        images = try container.decodeIfPresent([String].self, forKey: .images)
        rating = try container.decodeIfPresent(Double.self, forKey: .rating)
        tags = try container.decodeIfPresent([String].self, forKey: .tags)
        amapPoiId = try container.decodeIfPresent(String.self, forKey: .amapPoiId)
        distance = try container.decodeIfPresent(Double.self, forKey: .distance)
        type = try container.decodeIfPresent(String.self, forKey: .type)
        address = try container.decodeIfPresent(String.self, forKey: .address)
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(id, forKey: .id)
        try container.encode(name, forKey: .name)
        try container.encode(location, forKey: .location)
        try container.encodeIfPresent(category, forKey: .category)
        try container.encodeIfPresent(poiDescription, forKey: .poiDescription)
        try container.encodeIfPresent(images, forKey: .images)
        try container.encodeIfPresent(rating, forKey: .rating)
        try container.encodeIfPresent(tags, forKey: .tags)
        try container.encodeIfPresent(amapPoiId, forKey: .amapPoiId)
        try container.encodeIfPresent(distance, forKey: .distance)
        try container.encodeIfPresent(type, forKey: .type)
        try container.encodeIfPresent(address, forKey: .address)
    }
}

// MARK: - 路线卡片（用于列表展示）
struct RouteCard: Identifiable {
    let id: String
    let route: Route
    let cityName: String
    let likes: Int
    let views: Int
    let isFeatured: Bool
}

File: app/models/route.py
"""路线相关数据模型"""

from datetime import datetime
from typing import Any

from bson import ObjectId
from pydantic import BaseModel, Field, field_validator

from app.utils.helpers import PyObjectId


class Location(BaseModel):
    """地理位置坐标"""

    type: str = "Point"
    coordinates: list[float] = Field(..., min_length=2, max_length=2)  # [longitude, latitude]

    @field_validator("coordinates")
    @classmethod
    def validate_coordinates(cls, v: list[float]) -> list[float]:
        lon, lat = v[0], v[1]
        if not (-180 <= lon <= 180):
            raise ValueError("经度必须在 -180 到 180 之间")
        if not (-90 <= lat <= 90):
            raise ValueError("纬度必须在 -90 到 90 之间")
        return v


class RoutePointPhoto(BaseModel):
    """轨迹点照片（Base64内嵌）"""

    id: str = Field(default_factory=lambda: str(ObjectId()))
    data: str  # Base64编码的图片数据
    content_type: str = "image/jpeg"  # MIME类型
    caption: str | None = None  # 照片说明
    size: int = 0  # 原始文件大小(字节)
    created_at: datetime = Field(default_factory=datetime.now)


class RoutePoint(BaseModel):
    """路线上的一个点"""

    location: Location
    elevation: float | None = None  # 海拔高度(米)
    timestamp: datetime | None = None
    poi_id: str | None = None  # 关联的POI ID
    name: str | None = None
    description: str | None = None

    # 编辑功能扩展
    is_waypoint: bool = False  # 是否为用户标记的途径点
    photos: list[RoutePointPhoto] = Field(default_factory=list)  # 照片列表（最多5张）
    is_edited: bool = False  # 是否被用户编辑过
    original_location: Location | None = None  # 原始位置（编辑后保留）


class POI(BaseModel):
    """兴趣点(Point of Interest)"""

    id: str = Field(default_factory=lambda: str(ObjectId()))
    name: str
    location: Location
    category: str  # 景点、餐饮、休息点等
    description: str | None = None
    images: list[str] = Field(default_factory=list)
    rating: float | None = None
    tags: list[str] = Field(default_factory=list)
    amap_poi_id: str | None = None  # 高德地图POI ID


class Route(BaseModel):
    """路线模型"""

    id: str = Field(default_factory=lambda: str(ObjectId()), alias="_id")
    name: str
    description: str | None = None
    cover: str | None = None  # 封面图 URL

    # 路线数据
    points: list[RoutePoint] = Field(default_factory=list)
    pois: list[POI] = Field(default_factory=list)

    # 统计信息
    distance: float = 0.0  # 总距离(米)
    elevation_gain: float = 0.0  # 累计爬升(米)
    estimated_duration: int = 0  # 预计时间(分钟)

    # 位置信息
    start_location: Location  # 起点
    end_location: Location | None = None  # 终点
    city: str | None = None
    district: str | None = None

    # 交互数据
    favorites_count: int = 0
    views_count: int = 0
    completions_count: int = 0

    # 元数据
    difficulty: str = "medium"  # easy, medium, hard
    tags: list[str] = Field(default_factory=list)
    created_by: str | None = None
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)
    is_published: bool = True
    is_featured: bool = False  # 是否精选路线

    class Config:
        populate_by_name = True
        json_encoders = {ObjectId: str}


class GPSTrack(BaseModel):
    """GPS轨迹记录"""

    id: str = Field(default_factory=lambda: str(ObjectId()), alias="_id")
    route_id: str | None = None
    user_id: str | None = None

    # 轨迹点数据
    points: list[dict[str, Any]] = Field(default_factory=list)  # [{location, elevation, timestamp, ...}]

    # 统计信息
    distance: float = 0.0
    elevation_gain: float = 0.0
    duration: int = 0  # 实际用时(秒)
    average_speed: float = 0.0  # 平均速度(m/s)

    # 元数据
    started_at: datetime
    ended_at: datetime | None = None
    created_at: datetime = Field(default_factory=datetime.now)

    class Config:
        populate_by_name = True
        json_encoders = {ObjectId: str}
